### ISSUE TITLE
Failure in typing Twitter PIN causes new PIN to be generated

### ISSUE THREAD

**Issue by dcousineau** [spazproject/Spaz-Enyo#92]
If you fail to type in the correct PIN and attempt to log in (and get a failure), the previously correct PIN will no longer work (as I assume internally Twitter generates a new PIN). Thusly the PIN page must be refreshed and the new pin must be typed in proper.

When we integrate we will want to refresh the pin page on failure. Thanks to twitter for this one.

**Comment by willhoney7** [comment]
Hmmmm. To be honest, I don't think this would show up enough to warrant the amount of work needed to circumvent it. If we just display an error to the user when they enter the pin wrong saying, "Wrong pin, please refresh the page and get a new one", we'd be fine.

If we do the pin page as a separate browser webpage, I don't think we have control over it to refresh or even close it.

A webview is a possibility but would fit poorly with our popup I think.

A third option is to roll our own webpage card. We just pop open a new card that has our webview widget and that will allow us to refresh it and so forth. However, I don't think this will work in a browser environment.

**Comment by funkatron** [comment]
I might be able to sort this out in code; lemme look into it. It may be that I just am assuming the code is correct and never try to reset. maybe?

**Comment by dcousineau** [comment]
I personally would put in a vote to eventually move to using a webview in the workflow.

My main feature as per Tibfib is that I'm an advanced user and I spent 20 minutes wondering why I couldn't authenticate. If that happened to me, what happens to an average user who keeps typing in the same pin even though the pin changes every attempt? They'll get frustrated and move to another app. As well, what about the users with Dyslexia? They're probably going to have several attempts.

For the immediate term we could probably get away with a big warning message and a "Start Over" workflow.

**Comment by funkatron** [comment]
In 67878452068d455cf84335b844cc2e18448be7d2 I had it handle it more intelligently. The failure is detected and the user is encouraged to "Try Again." Only issue, really, is that I can't reach out and tell the browser to close it's window from before. I think this works decently enough for now, but I suspect a webview implementation in-app is going to be something we want to do post initial release. For now we have higher priorities, I think.

**Comment by willhoney7** [comment]
How would a webview work with our popup workflow? 

Why not just have the banner error for the few power users say, "Wrong pin, refresh Twitter and get a new one."

**Comment by funkatron** [comment]
It would open the webview within the popup, probably.

Power users aren't our worry here. Trust me, as the guy who has to answer questions from users.

## 

Ed Finkler
http://funkatron.com
@funkatron
AIM: funka7ron / ICQ: 3922133 / XMPP:funkatron@gmail.com

**Comment by dcousineau** [comment]
That was exactly my point: Power users aren't going to have this issue (though they may, like I did).

We could probably send a flag / alter the URL and force the mobile interface of the Twitter PIN auth so it'll fit in the dialog.

**Comment by funkatron** [comment]
Yep, forcing mobile would be a really good idea.Unrelated, from a friend today:"Yo -- finally giving the "chrome" version of spaz a try. Best. Linux. Client. Ever. :)"Ed Finkler@funkatronFunkatron.comSent from my Palm PreOn Jun 1, 2011 5:17 PM, dcousineau reply@reply.github.com wrote: That was exactly my point: Power users aren't going to have this issue (though they may, like I did).

## We could probably send a flag / alter the URL and force the mobile interface of the Twitter PIN auth so it'll fit in the dialog.

Reply to this email directly or view it on GitHub:
https://github.com/funkatron/Spaz-Enyo/issues/92#comment_1277223

**Comment by willhoney7** [comment]
Oh, looks my email reply didn't get added. Here it was: 

Oh. I thought your point was that power users would have issues since you did. 

If we do a webview, then why not go the whole way and do the redirect stuff?

**Comment by funkatron** [comment]
Yeah, that won't work because we can't redirect to anything, basically. Capturing the redirect event seems flakey. Rather go with known entity for now so it works.

**Comment by willhoney7** [comment]
Gotcha, sounds good.
